File: Update.fs
module Update 

open System
open System.Text

open Settings
open FSharpHttpRequest
open FSharpTools
open FSharpTools.Functional
open AsyncResult
open FSharpTools.Async

// TODO to FSharpHttpRequest
let basicAuthentication name passwd = 
    {
        Key = "Authorization"
        Value = "Basic " + Convert.ToBase64String (Encoding.UTF8.GetBytes (name + ":" + passwd))
    }

let update settings host ip = 

    let getUrl () = sprintf "https://%s/nic/update?hostname=%s&myip=%s" settings.Provider host ip

    let mapResult (resStr: string) = 
        printfn "response: %s" resStr 
        () 
        |> Async.toAsync

    let printUrl url = printfn "Updating %s" url

    match DnsCheck.check host ip with
    | false ->
        Request.getString { 
            Request.defaultSettings with 
                Url = (getUrl () ) |> sideEffect printUrl
                Headers = Some [| 
                    { Key = "User-Agent"; Value = "DynDNS Updater" }
                    basicAuthentication settings.Account settings.Passwd
                |]  
            }
            |> AsyncResult.mapError Error.fromUpdateResult
            |>> mapResult

    | true ->
        printfn "%s: IP Address not changed, no action needed\n" host
        Ok () 
        |> toAsync

File: Update.cs
using CsTools.Extensions;
using CsTools.Functional;
using CsTools.HttpRequest;

using static System.Console;
using static CsTools.Core;
using static CsTools.HttpRequest.Core;

static class Update 
{
    public static Task<string> Run(Settings settings, string ip, string domain)
        => RepeatOnError(async () =>
        {
            if (Dns.Check(domain, ip))
            {
                WriteLine($"{domain}: IP Address not changed, no action needed");
                return "";
            }
            else
                return await UpdateRequest(settings, ip, domain);
        }, new RepeatOnErrorOptions() {RepeatCount = 4, WaitTime = TimeSpan.FromSeconds(5) });
            
    static async Task<string> UpdateRequest(Settings settings, string ip, string domain)
    {
        var response = await Request.RunAsync(DefaultSettings with
        {
            Method = HttpMethod.Get,
            BaseUrl = $"https://{settings.Provider}",
            Headers = [
                        new Header("User-Agent", "DynDNS Updater"),
                        BasicAuthentication.From(settings.Account, settings.Passwd)],
            Url = $"/nic/update?hostname={domain.SideEffect(d => WriteLine($"Updating {d}"))}&myip={ip}"
        });
        var result = await response.Content.ReadAsStringAsync();
        WriteLine($"Response: {result}");
        return result;
    }
}
